src/powerquery-formatter/themes/themes.ts

```diff
@@ -0,0 +1,374 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as PQP from "@microsoft/powerquery-parser";
+
+import { IParameters, IRawTheme, IRawThemeSetting } from "./types";
+import { strArrCmp, strcmp } from "./utils";
+
+/**
+ * An initially parsed rule of a scopeList from raw theme
+ *  a scope list would be like:
+ *    "scope1 scope2 scope3 scope4"
+ */
+export class ParsedThemeRule<T extends IParameters = IParameters> {
```

Comment:
Does this get inherited / extended anywhere, or is this just a data class?

Reply:
Just a data class for now, like DAO, i suggest we leave it as a class other than making it a plain js object. I was worrying that we need to pre-prepare some fancy scope-selectors data-structures or store some ad-hoc parameters which may require additional validation on IRawTheme, by then this ParsedThemeRule class would be the ideal place to put some utils or helper methods relating to the Theme data massage. 😬 Let's leave it OOP inside theme module, which won't harm any.